overlord/patch/patch4_test.go

```diff
@@ -0,0 +1,305 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2016 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package patch_test
+
+import (
+	"io/ioutil"
+	"os"
+	"path/filepath"
+
+	. "gopkg.in/check.v1"
+
+	"github.com/snapcore/snapd/dirs"
+	"github.com/snapcore/snapd/overlord/patch"
+	"github.com/snapcore/snapd/overlord/snapstate"
+	"github.com/snapcore/snapd/overlord/state"
+)
+
+type patch4Suite struct{}
+
+var _ = Suite(&patch4Suite{})
+
+var statePatch4JSON = []byte(`
+{
+	"last-task-id": 999,
+	"last-change-id": 99,
+
+	"data": {
+		"patch-level": 3,
+		"snaps": {
+			"a": {
+				"sequence": [
+					{"name": "", "revision": "1"},
+					{"name": "", "revision": "2"},
+					{"name": "", "revision": "3"}],
+				"current": "2"},
+			"b": {
+				"sequence": [
+					{"name": "", "revision": "1"},
+					{"name": "", "revision": "2"}],
+				"current": "2"}
+		}
+	},
+	"changes": {
+		"1": {
+			"id": "1",
+			"kind": "revert-snap",
+			"summary": "revert a snap",
+			"status": 0,
+			"data": {"snap-names": ["a"]},
+			"task-ids": ["1","2","3","4"]
+		},
+		"2": {
+			"id": "2",
+			"kind": "refresh-snap",
+			"summary": "refresh b snap",
+			"status": 0,
+			"data": {"snap-names": ["b"]},
+			"task-ids": ["10","11","12","13","14","15","16"]
+		}
+	},
+	"tasks": {
+		"1": {
+			"id": "1",
+			"kind": "prepare-snap",
+			"summary": "",
+			"status": 4,
+			"data": {
+			    "snap-setup": {
+				"side-info": {"revision": "2", "name": "a"}
+			    }
+			},
+			"halt-tasks": ["2"],
+			"change": "1"
+		},
+		"2": {
+			"id": "2",
+			"kind": "unlink-current-snap",
+			"summary": "",
+			"status": 4,
+			"data": {
+			    "snap-setup-task": "1"
+			},
+			"wait-tasks": ["1"],
+			"halt-tasks": ["3"],
+			"change": "1"
+		},
+		"3": {
+			"id": "3",
+			"kind": "setup-profiles",
+			"summary": "",
+			"status": 4,
+			"data": {
+			    "snap-setup-task": "1"
+			},
+			"wait-tasks": ["2"],
+			"halt-tasks": ["4"],
+			"change": "1"
+		},
+		"4": {
+			"id": "4",
+			"kind": "link-snap",
+			"summary": "make snap avaiblabla",
+			"status": 4,
+			"data": {
+			    "had-candidate": true,
+			    "snap-setup-task": "1"
+			},
+			"wait-tasks": ["3"],
+			"change": "1"
+		},
+
+		"10": {
+			"id": "10",
+			"kind": "download-snap",
+			"summary": "... download ...",
+			"status": 4,
+			"data": {"snap-setup": {"side-info": {"revision": "2", "name": "a"}}},
```

Comment:
I guess its test data and it does not matter but I could not help noticing that 2 is already in the index in the example data.